### ISSUE TITLE
Update EventStream to Property binding to remove glitchiness

### ISSUE THREAD

**Issue by mikebaum** [mikebaum/RxUI#34]
A glitch in an FRP system is described well in the following [link](http://stackoverflow.com/questions/25139257/terminology-what-is-a-glitch-in-functional-reactive-programming-rx) and futher [here](glitchiness-and-redundant-computation).

In a nutshell a glitch is a temporary inconsistent state that happends because event propagation is not instantaneous. Consider the following:

``` java
EventSubject<Integer> integerSubject = EventSubject.create();
Property<Integer> integerProperty1 = Property.create(0);
Property<Integer> integerProperty2 = Property.create(0);

PropertyStream<Integer> int1PlusInt2 =
    Property.combine(integerProperty1, integerPropety2, (a,b) -> a + b);

integerSubject.onEvent(integerProperty1::setValue);
integerSubject.onEvent(integerProperty2::setValue);

int1PlusIn2.onChanged(System.out::println);

integerSubject.publish(2);
```

Without glitches the expected output of this code would be:

```
4
```

However, due to glitches, the actual output is:

```
2 <-- glitch
4
```

In this case, the first output of 2 would be a glitch. It was observable only due to the fact that event propagation is not instantaneous, since subscribers are called back in sequence.

**Comment by mikebaum** [comment]
In order to eliminate glitches there are a few different approaches that can be taken. As described in [Glitchiness and Redundant Computation](https://github.com/lihaoyi/scala.rx#glitchiness-and-redundant-computation) two approaches would be:
1. Make the dataflow graph static, and perform a topological sort to rank nodes in the order they are to be updated. This means that a node always is updated after its dependencies, meaning they will never see any stale data
2. Pause the updating of a node when it tries to call upon a dependency which has not been updated. This could be done by blocking the thread, for example, and only resuming after the dependency has been updated.

At this point (ignoring EventStreams) Properties and Property Streams are glitch free. This is due to the fact that all property subscribers are sorted to process bindings first. This means that when a property is updated all other properties that are bound to that property are updated before any other subscribers are notified. This combined with the fact that all properties (and transformed properties by any of the operators that produce a property stream) offer a getter that always provides the current value means that no observer of a property can observe an inconsistent state. Also, the fact that all property operators should guarantee to suppress duplicate consecutive updates all but assures that properties (in exclusion of EventStreams) are glitch free.

In order to extend this glitch prevention to systems that include both properties and event streams a similar treatment will need to be applied to the event stream --> property bindings. The solution will be to use approach 2 as described above. Therefore all event stream --> property bindings will be processed first when propagating an event from an event stream. When a property is updated via an event stream, it's propagation will be paused until the event stream that caused the update has completed it's dispatch. This should assure that all properties have been updated before dispatching to their observers and therefore the description of glitch free propagation as described in the previous section would apply.

If these two techniques for glitch prevention are to hold, it is essential that properties are only allowed to be bound to other properties and/or streams via the method `Property.bind()`. It will therefore be necessary to block updates to properties via the following two method calls:

```
eventStream.onEvent(property::setValue);
property.onChanged(property::setValue)
```

The reason for this is that we will not be able to determine up front that these two method calls actually represent a binding and hence the protective measures will be avoided.

**Comment by mikebaum** [comment]
closed as per PR #35.
